internal/chain/engine.go

```diff
@@ -273,6 +265,25 @@ func (e *engineImpl) Finalize(
 	isBeaconChain := header.ShardID() == shard.BeaconChainShardID
 	isNewEpoch := len(header.ShardState()) > 0
 	inStakingEra := chain.Config().IsStaking(header.Epoch())
+	if isBeaconChain && inStakingEra {
```

Comment:
These code can be packed into state.AddReward and stay inside the AccumulateReward func. Doesn't have to be another separate loop update.